dotnet-desktop-guide/net/wpf/events/routed-events-overview.md

```diff
@@ -0,0 +1,260 @@
+
+- From an **implementation** perspective, a routed event is an event registered with the WPF event system, backed by an instance of the <xref:System.Windows.RoutedEvent> class, and processed by the WPF event system. Typically, a routed event is implemented with a CLR event "wrapper" to enable attaching handlers in XAML and in code-behind as you would a CLR event.
+
+WPF applications typically contain many elements, which were either declared in XAML or instantiated in code. An application's elements exists within its element tree. Depending on how a routed event is defined, when the event is raised on a source element it:
+
+- Bubbles up through element tree from the source element to the root element, which is typically a page or window.
+- Tunnels down through the element tree from the root element to the source element.
+- Doesn't travel through the element tree, and only occurs on the source element.
+
+Consider the following partial element tree:
+
+:::code language="xaml" source="./snippets/routed-events-overview/csharp/MainWindow.xaml" id="YesNoCancelButtons":::
+
+The element tree renders as shown:
+
+:::image type="content" source="./media/routed-events-overview/yes-no-cancel.png" border="true" alt-text="Yes, No, and Cancel buttons.":::
+
+Each of the three buttons is a potential <xref:System.Windows.Controls.Primitives.ButtonBase.Click> event source. When one of the buttons is clicked, it raises the `Click` event that bubbles up from the button to the root element. The <xref:System.Windows.Controls.Button> and <xref:System.Windows.Controls.Border> elements don't have event handlers attached, but the <xref:System.Windows.Controls.StackPanel> does. Possibly other elements higher up in the tree that aren't shown also have `Click` event handlers attached. When the `Click` event reaches the `StackPanel` element, the WPF event system invokes the `YesNoCancelButton_Click` handler. The event route for the `Click` event in the example is: `Button` -> `StackPanel` -> `Border` -> successive parent elements.
+
+### Top-level scenarios for routed events
+
+Here are some of the scenarios that motivated the routed event concept, and distinguish it from a typical CLR event:
+
+- **Control composition and encapsulation**: Various controls in WPF have a rich content model. For example, you can place an image inside a <xref:System.Windows.Controls.Button>, which effectively extends the visual tree of the button. But, the added image mustn't break the hit-test behavior of the button, which needs to respond when a user clicks the image pixels.
+
+- **Singular handler attachment points**: In Windows Forms, you have to attach the same handler multiple times to process events raised on multiple elements. Routed events enable you to attach that handler only once, as shown in the previous XAML example. Event handler logic can determine where the event came from. For example, this handler is applicable to the previously shown XAML element tree:
```

Comment:
I don't much like this. Back when WPF came out, the only other managed UI stack was Windows Forms. You had a lot of people converting from it, so a statement like this made sense. It no longer makes sense to assume people know what WinForms is or how it even works. I tried to edit this to make the benefit clearer. How about this?

```suggestion
- **Singular handler attachment points**: You could register a handler with each button's `Click` event, but, routed events enables you to attach a single handler. This enables you to change element tree under the singular handler, such as adding or removing more buttons, without having to register each button's `Click` event. When the `Click` event is raised, handler logic can determine where the event came from. For example, this handler is applicable to the previously shown XAML element tree:
```

Reply:
Looks good, just added in a reference to the previous XAML example.